File: UniBlock/src/main/scala/service/Miner.scala
package it.unifi.nave.uniblock
package service

import data.block.BlockHeader

import java.util.concurrent.Callable

case class Miner(blockHeader: BlockHeader, start: Int, end: Int) extends Callable[Option[Int]] {
  private var interrupted = false

  override def call(): Option[Int] =
    LazyList.range(start, end).map(setNonce).filterNot(_ => isInterrupted).find(_.isMined).map(_.nonce)

  private def setNonce(nonce: Int): BlockHeader = {
    blockHeader.nonce = nonce
    blockHeader
  }

  private def isInterrupted: Boolean = {
    interrupted = Thread.interrupted()
    interrupted
  }
}

File: program/src/main/java/it/unifi/nave/uniblock/service/Miner.java
package it.unifi.nave.uniblock.service;

import it.unifi.nave.uniblock.data.block.BlockHeader;

import java.util.Optional;
import java.util.concurrent.Callable;

public record Miner(BlockHeader blockHeader, int start, int end)
    implements Callable<Optional<Integer>> {
  @Override
  public Optional<Integer> call() {
    Integer result = null;
    for (int i = start; i < end && !Thread.interrupted(); i++) {
      blockHeader.setNonce(i);
      if (blockHeader.isMined()) {
        result = i;
        break;
      }
    }
    return Optional.ofNullable(result);
  }
}
